sdk/communication/azure-communication-callingserver/azure/communication/callingserver/_download.py

```diff
@@ -0,0 +1,358 @@
+        return self._download_chunk(chunk_start, chunk_end - 1)
+
+    def _update_progress(self, length):
+        if self.progress_lock:
+            with self.progress_lock:  # pylint: disable=not-context-manager
+                self.progress_total += length
+        else:
+            self.progress_total += length
+
+    def _write_to_stream(self, chunk_data, chunk_start):
+        if self.stream_lock:
+            with self.stream_lock:  # pylint: disable=not-context-manager
+                self.stream.seek(self.stream_start +
+                                 (chunk_start - self.start_index))
+                self.stream.write(chunk_data)
+        else:
+            self.stream.write(chunk_data)
+
+    def _download_chunk(self, chunk_start, chunk_end):
+        range_header = CallingServerUtils.validate_and_format_range_headers(
+            chunk_start,
+            chunk_end
+        )
+
+        response = self.client.download(
+            content_url=self.endpoint,
+            http_range=range_header,
+            **self.request_options
+        )
+
+        #pylint: disable=protected-access
+        return response.response.internal_response.content
+
+
+class _ChunkIterator(object):
+    """sync iterator for chunks in content download stream."""
+
+    def __init__(self, size, content, downloader, chunk_size):
+        self.size = size
+        self._chunk_size = chunk_size
+        self._current_content = content
+        self._iter_downloader = downloader
+        self._iter_chunks = None
+        self._complete = (size == 0)
+
+    def __len__(self):
+        return self.size
+
+    def __iter__(self):
+        raise TypeError("stream must be iterated synchronously.")
+
+    def __aiter__(self):
+        return self
+
+    def __anext__(self):
+        """Iterate through responses."""
+        if self._complete:
+            raise StopIteration("Download complete")
+        if not self._iter_downloader:
+            # cut the data obtained from initial GET into chunks
+            if len(self._current_content) > self._chunk_size:
+                return self._get_chunk_data()
+            self._complete = True
+            return self._current_content
+
+        if not self._iter_chunks:
+            self._iter_chunks = self._iter_downloader.get_chunk_offsets()
+
+        # initial GET result still has more than _chunk_size bytes of data
+        if len(self._current_content) >= self._chunk_size:
+            return self._get_chunk_data()
+
+        try:
+            chunk = next(self._iter_chunks)
+            self._current_content += self._iter_downloader.yield_chunk(chunk)
+        except StopIteration:
+            self._complete = True
+            # it's likely that there some data left in self._current_content
+            if self._current_content:
+                return self._current_content
+            # https://stackoverflow.com/questions/27318327/how-to-imitate-python-3s-raise-from-in-python-2
```

Comment:
Is this ok to have in an SDK?

Reply:
Links to stackoverflow articles

Reply:
If you search within the SDK you'll see that numerous references to stackoverflow, GitHub and other similar sites are mentioned. 
Anyway, I don't see why to keep these ones.